modules/compute/src/main/java/org/apache/ignite/compute/internal/IgniteComputeImpl.java

```diff
@@ -0,0 +1,302 @@
+        this.clusterService = clusterService;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public <T, R, F> F execute(Class<? extends ComputeTask<T, R, F>> clazz, T argument) throws ComputeException {
+        UUID executionId = UUID.randomUUID();
+
+        Execution<R> execution = registerExecution(clazz, argument, executionId);
+
+        Collection<ExecutionResult<R>> executionResults = execution.getResult();
+
+        executions.remove(executionId);
+
+        ComputeTask<T, R, F> computeTask;
+        try {
+            Constructor<? extends ComputeTask<T, R, F>> declaredConstructor = clazz.getDeclaredConstructor();
+            declaredConstructor.setAccessible(true);
+            computeTask = declaredConstructor.newInstance();
+        } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+            throw new ComputeException("Failed to execute reduce: " + e.getMessage(), e);
+        }
+
+        return computeTask.reduce(executionResults.stream().map(ExecutionResult::result).collect(toList()));
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void start() {
+        clusterService.messagingService().addMessageHandler(ComputeMessagesType.class, new ComputeNetworkHandler());
+
+        clusterService.topologyService().addEventHandler(new TopologyHandler());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void stop() throws Exception {
+        IgniteUtils.shutdownAndAwaitTermination(computeExecutor, 5, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Creates new execution based on a task, an argument and an id.
+     *
+     * @param clazz Task's class.
+     * @param argument Task's argument.
+     * @param executionId Execution id.
+     * @param <T> Argument's type.
+     * @param <R> Intermediate result's type.
+     * @param <F> Result's type.
+     * @return Execution.
+     */
+    @NotNull
+    private <T, R, F> Execution<R> registerExecution(Class<? extends ComputeTask<T, R, F>> clazz, T argument, UUID executionId) {
+        Collection<ClusterNode> currentTopology = this.clusterService.topologyService().allMembers();
+
+        List<String> nodeIds = currentTopology.stream().map(ClusterNode::id).collect(toList());
+
+        var execution = new Execution<R>(nodeIds);
+
+        Execution<?> prev = executions.putIfAbsent(executionId, execution);
+        assert prev == null;
+
+        MessagingService messagingService = this.clusterService.messagingService();
+
+        ExecuteMessage executeMessage = messagesFactory.executeMessage()
+                .executionId(executionId).className(clazz.getName()).argument(argument).build();
+
+        currentTopology.forEach(clusterNode -> {
+            try {
+                messagingService.send(clusterNode, executeMessage).get(3, TimeUnit.SECONDS);
+            } catch (InterruptedException | ExecutionException | TimeoutException e) {
```

Comment:
What if a different exception is thrown here? Will the execution get stuck forever?

Reply:
Not sure if Future#get can throw other exceptions.

Reply:
`MessagingService#send()` might throw anything :)

Reply:
It has no `@throws` as I see

Reply:
How about separating try/catch block into two: one for `#send()` (which is our own and can throw anything, like some `RuntimeException` due to a programming error), where we'd try to make sure that the execution never gets stuck even if some `IllegalStateException` is thrown, and the second one, for `.get()` (to which we can... hmmm... trust)?